include/date/tz.h

```diff
@@ -2183,42 +2197,224 @@ from_stream(std::basic_istream<CharT, Traits>& is, const CharT* fmt,
-to_sys_time(const gps_time<Duration>& t)
+template<>
+struct clock_time_conversion<utc_clock, utc_clock>
 {
-    return to_sys_time(to_utc_time(t));
-}
+  template <class Duration>
+  auto operator()(const std::chrono::time_point<utc_clock, Duration>& st)
+    -> std::chrono::time_point<utc_clock, Duration>
+  { 
+    return st;
+  }
+};
 
-template <class Duration>
-inline
-tai_time<typename std::common_type<Duration, std::chrono::seconds>::type>
-to_tai_time(const gps_time<Duration>& t) NOEXCEPT
+template<>
+struct clock_time_conversion<std::chrono::system_clock, utc_clock>
 {
-    using namespace std::chrono;
-    using duration = typename std::common_type<Duration, seconds>::type;
-    return tai_time<duration>{t.time_since_epoch()} +
-            (sys_days(year{1980}/jan/sun[1]) - sys_days(year{1958}/jan/1) + seconds{19});
-}
+  template <class Duration>
+  utc_time<typename std::common_type<Duration, std::chrono::seconds>::type>
+  operator()(const sys_time<Duration>& st) const
+  {
+     return utc_clock::from_sys(st);
+  }
+};
 
-template <class Duration>
-inline
-gps_time<typename std::common_type<Duration, std::chrono::seconds>::type>
-to_gps_time(const tai_time<Duration>& t) NOEXCEPT
+template<>
+struct clock_time_conversion<utc_clock, std::chrono::system_clock>
 {
-    using namespace std::chrono;
-    using duration = typename std::common_type<Duration, seconds>::type;
-    return gps_time<duration>{t.time_since_epoch()} -
-            (sys_days(year{1980}/jan/sun[1]) - sys_days(year{1958}/jan/1) + seconds{19});
+  template <class Duration>
+  sys_time<typename std::common_type<Duration, std::chrono::seconds>::type>
+  operator()(const utc_time<Duration>& ut) const
+  {
+    return utc_clock::to_sys(ut);
+  }
+};
+
+template<typename Clock>
+struct clock_time_conversion<Clock, Clock>
+{
+  template <class Duration>
+  auto operator()(const std::chrono::time_point<Clock, Duration>& st)
+    -> std::chrono::time_point<Clock, Duration>
+  { 
+    return st;
+  }
+};
+
+namespace ctc_detail
+{
+  //Check if TimePoint is time for given clock, 
+  //if so exposes it as type typedef
+  template<typename Clock, typename TimePoint>
+  struct return_clock_time
+    : std::enable_if<
+        std::is_same<Clock, typename TimePoint::clock>::value,
+        TimePoint
+      >
+  {};
+
+  // Check if Clock has to_sys method accepting TimePoint with given duration const& and returning sys_time
+  // If so has nested type member equal to return type to_sys.
+  template<typename Clock, typename Duration, typename = void>
+  struct return_to_sys 
+  {};
+
+  template<typename Clock, typename Duration>
+  struct return_to_sys<Clock, Duration, decltype(Clock::to_sys(std::declval<std::chrono::time_point<Clock, Duration> const&>()), void())>
+    : return_clock_time<std::chrono::system_clock, typename std::decay<decltype(Clock::to_sys(std::declval<std::chrono::time_point<Clock, Duration> const&>()))>::type>
```

Comment:
Can we get rid of the `decay` trait here?  I can't imagine `to_sys` would not return by value, and if it did, it might be good to catch it as an error.

Reply:
Just removing decay would lead to eliminating overload, which does not necessary means producing an error - example may select `utc_clock` conversion path, like in case of `amb1_clock` example.
The alternative would be to put an ``static_assert`` in `return_clock_time` that would check if TimePoint is not an reference.

Reply:
I like the idea. I will change `return_clock_time<C, TP>` to include an `static_assert` on `TP` being the same as `std::chrono::time_point<C, typename TP::duration>`. So havin `to_sys` returning something different than `sys_time` will be hard error, instead of SFINAE.

Reply:
Updated in next commit.